cmds/jfrogplugin/uploaders/helm/helm.go

```diff
@@ -0,0 +1,248 @@
+	}
+
+	info, err := ConvertTargetSpecToInfo(targetSpec)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert target spec to info: %w", err)
+	}
+
+	return info, nil
+}
+
+// Upload uploads any artifact that is of type artifacttypes.HELM_CHART.
+// Process:
+//  1. introspect the JFrogHelmUploaderSpec (cast from ppi.UploadTargetSpec) and hint parameter
+//     (the hint is expected to be an OCI style reference, such as `repo/comp:version`)
+//  2. building an Artifactory Style JFrog Upload URL out of it (see ConvertTargetSpecToHelmUploadURL),
+//  3. creating a request respecting the passed credentials based on SetHeadersFromCredentials
+//  4. uploading the passed blob as is (expected to be a tgz byte stream)
+//  5. intepreting the JFrog API response, and converting it from ArtifactoryUploadResponse to ppi.AccessSpec
+func (a *Uploader) Upload(_ ppi.Plugin, artifactType, _, hint string, targetSpec ppi.UploadTargetSpec, creds credentials.Credentials, reader io.Reader) (ppi.AccessSpecProvider, error) {
+	if artifactType != artifacttypes.HELM_CHART {
+		return nil, fmt.Errorf("unsupported artifact type %s", artifactType)
+	}
+
+	spec, ok := targetSpec.(*JFrogHelmUploaderSpec)
+	if !ok {
+		return nil, fmt.Errorf("the type %T is not a valid target spec type", spec)
+	}
+
+	if err := EnsureSpecWithHelpFromHint(spec, hint); err != nil {
+		return nil, fmt.Errorf("could not ensure spec to be ready for upload: %w", err)
+	}
+
+	targetURL, err := ConvertTargetSpecToHelmUploadURL(spec)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert target spec to URL: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.TODO(), spec.GetTimeout())
+	defer cancel()
+
+	access, err := Upload(ctx, reader, a.Client, targetURL, creds)
+	if err != nil {
+		return nil, fmt.Errorf("failed to upload")
+	}
+
+	return func() ppi.AccessSpec {
+		return access
+	}, nil
+}
+
+// ConvertTargetSpecToInfo converts the JFrogHelmUploaderSpec
+// to a valid info block containing the consumer ID used
+// in the library to identify the correct credentials that need to
+// be passed to it.
+func ConvertTargetSpecToInfo(spec *JFrogHelmUploaderSpec) (*ppi.UploadTargetSpecInfo, error) {
+	purl, err := parseURLAllowNoScheme(spec.URL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse URL: %w", err)
+	}
+
+	var info ppi.UploadTargetSpecInfo
+
+	// By default, we identify an artifactory repository as a combination
+	// of Host & Repository
+	info.ConsumerId = credentials.ConsumerIdentity{
+		cpi.ID_TYPE:   NAME,
+		ID_HOSTNAME:   purl.Hostname(),
+		ID_REPOSITORY: spec.Repository,
+	}
+	if purl.Port() != "" {
+		info.ConsumerId.SetNonEmptyValue(ID_PORT, purl.Port())
+	}
+
+	return &info, nil
+}
+
+// EnsureSpecWithHelpFromHint introspects the hint and fills the target spec based on it.
+// It makes sure that the spec can be used to access a JFrog Artifactory HELM Repository.
+func EnsureSpecWithHelpFromHint(spec *JFrogHelmUploaderSpec, hint string) error {
```

Comment:
So I was wondering. Why not make the thing that you want part of the spec instead of using the hint as an extra information? Like, it could be part of the `Config` you are constructing, unless I completely misunderstand how the plug works. 🤔

Reply:
Unfortunately that doesnt really work for dynamic repos. 

Assume you want to have a config like in the example I provided in the PR description. In this case the repo name and chart name + version are determined dynamically. If you would put them into the spec instead of the hint, every single helm chart would get transferred to the name+version that is specified in the spec. The hint however is dynamic, and can change for every single artifact.

Reply:
Ah, I get it. Thanks for the explanation.